### PR TITLE
Support deep updates

### PR DESCRIPTION
**Pull request by mikepizzo**
-Added support for Reading/Writing delta content in single resource request payloads
-Added tests for reading/writing single request resource with nested delta content
-Disabled (by default) reading/writing single response resource with nested delta content

### Description

This Pull Request enables reading/writing 4.01 deep update requests.

### Checklist (Uncheck if it is not completed)

- [x] *Test cases added*
- [x] *Build and test with one-click build and test script passed*

### CODE REVIEW

**Comment by mikepizzo** [comment]
Looks like there are still a few failures:

* src\Microsoft.OData.Client\Serialization\Serializer.cs(591,69): Error CS0117: 'Strings' does not contain a definition for 'Serializer_NullCollectionParameterItemValue'
* src\Microsoft.OData.Client\Serialization\Serializer.cs(600,65): Error CS0117: 'Strings' does not contain a definition for 'Serializer_InvalidCollectionParameterItemType'
*src\Microsoft.OData.Client\Serialization\Serializer.cs(652,69): Error CS0117: 'Strings' does not contain a definition for 'Serializer_InvalidCollectionParameterItemType'
* src\Microsoft.OData.Client\Serialization\Serializer.cs(898,61): Error CS0117: 'Strings' does not contain a definition for 'Serializer_InvalidCollectionParameterItemType'
* test\FunctionalTests\Microsoft.OData.Core.Tests\AnnotationFilterTests.cs(52,43): Error CS0117: 'AnnotationFilter' does not contain a definition for 'CreateInclueAllFilter'

File: src/Microsoft.OData.Core/ODataWriterCore.cs [hunk 1]
```diff
@@ -67,7 +67,7 @@ internal abstract class ODataWriterCore : ODataWriter, IODataOutputInStreamError
 
             this.outputContext = outputContext;
             this.writingResourceSet = writingResourceSet;
-            this.writingDelta = writingDelta;
+            this.writingDelta = writingDelta || (!writingResourceSet && !outputContext.WritingResponse);
```
Comment (xuzhg) [inline src/Microsoft.OData.Core/ODataWriterCore.cs]: >(!writingResourceSet && !outputContext.WritingResponse) [](start = 48, length = 55)

Can u illustrate the change here?

File: src/Microsoft.OData.Core/ODataWriter.cs [hunk 2]
```diff
@@ -51,7 +51,10 @@ public ODataWriter Write(ODataResourceSet resourceSet, Action nestedAction)
         /// <summary> Asynchronously start writing a resource set. </summary>
         /// <returns>A task instance that represents the asynchronous write operation.</returns>
         /// <param name="resourceSet">The resource set or collection to write.</param>
-        public abstract Task WriteStartAsync(ODataResourceSet resourceSet);
+        public virtual Task WriteStartAsync(ODataResourceSet resourceSet)
+        {
+            return TaskUtils.GetTaskForSynchronousOperation(() => this.WriteStart(resourceSet));
+        }
```
Comment (xuzhg) [inline src/Microsoft.OData.Core/ODataWriter.cs]: all the changes in this class looks not related to Deep updates? Right? 

However, i am ok to make these changes.

File: src/Microsoft.OData.Core/ODataWriterCore.cs (hunk 1 above)
Comment (mikepizzo) [reply src/Microsoft.OData.Core/ODataWriterCore.cs]: delta is valid when writing any top-level resource in a request.

---
In reply to: [348131350](https://github.com/OData/odata.net/pull/1585#discussion_r348131350) [](ancestors = 348131350)

File: src/Microsoft.OData.Core/ODataWriter.cs (hunk 2 above)
Comment (mikepizzo) [reply src/Microsoft.OData.Core/ODataWriter.cs]: Right; this was the separate issue to support async. I added it because I was in the code, but you're right, this is not related to deep updates.

---
In reply to: [348191583](https://github.com/OData/odata.net/pull/1585#discussion_r348191583) [](ancestors = 348191583)

**Comment by mikepizzo** [comment]
Will do; thanks!

---
In reply to: [319383510](https://github.com/OData/odata.net/pull/1585#issuecomment-319383510) [](ancestors = 319383510)
